dmriprep/interfaces/bids.py

```diff
@@ -0,0 +1,139 @@
+    >>> res = dsink.run()
+    >>> res.outputs.out_file  # doctest: +ELLIPSIS
+    '.../dmriprep/sub-02/ses-noanat/func/sub-02_ses-noanat_task-rest_run-01_\
+desc-preproc_bold.nii.gz'
+    """
+    input_spec = DerivativesDataSinkInputSpec
+    output_spec = DerivativesDataSinkOutputSpec
+    out_path_base = "dmriprep"
+    _always_run = True
+
+    def __init__(self, out_path_base=None, **inputs):
+        super(DerivativesDataSink, self).__init__(**inputs)
+        self._results['out_file'] = []
+        if out_path_base:
+            self.out_path_base = out_path_base
+
+    def _run_interface(self, runtime):
+        src_fname, _ = _splitext(self.inputs.source_file)
+        src_fname, dtype = src_fname.rsplit('_', 1)
+        _, ext = _splitext(self.inputs.in_file[0])
+        if self.inputs.compress is True and not ext.endswith('.gz'):
+            ext += '.gz'
+        elif self.inputs.compress is False and ext.endswith('.gz'):
+            ext = ext[:-3]
+
+        m = BIDS_NAME.search(src_fname)
+
+        mod = os.path.basename(os.path.dirname(self.inputs.source_file))
+
+        base_directory = runtime.cwd
+        if isdefined(self.inputs.base_directory):
+            base_directory = str(self.inputs.base_directory)
+
+        out_path = '{}/{subject_id}'.format(self.out_path_base, **m.groupdict())
+        if m.groupdict().get('session_id') is not None:
+            out_path += '/{session_id}'.format(**m.groupdict())
+        out_path += '/{}'.format(mod)
+
+        out_path = os.path.join(base_directory, out_path)
+
+        os.makedirs(out_path, exist_ok=True)
+
+        if isdefined(self.inputs.prefix):
+            base_fname = os.path.join(out_path, self.inputs.prefix)
+        else:
+            base_fname = os.path.join(out_path, src_fname)
+
+        formatstr = '{bname}{space}{desc}{suffix}{dtype}{ext}'
+        if len(self.inputs.in_file) > 1 and not isdefined(self.inputs.extra_values):
+            formatstr = '{bname}{space}{desc}{suffix}{i:04d}{dtype}{ext}'
+
+        space = '_space-{}'.format(self.inputs.space) if self.inputs.space else ''
+        desc = '_desc-{}'.format(self.inputs.desc) if self.inputs.desc else ''
+        suffix = '_{}'.format(self.inputs.suffix) if self.inputs.suffix else ''
+        dtype = '' if not self.inputs.keep_dtype else ('_%s' % dtype)
+
+        self._results['compression'] = []
+        for i, fname in enumerate(self.inputs.in_file):
+            out_file = formatstr.format(
+                bname=base_fname,
+                space=space,
+                desc=desc,
+                suffix=suffix,
+                i=i,
+                dtype=dtype,
+                ext=ext)
+            if isdefined(self.inputs.extra_values):
+                out_file = out_file.format(extra_value=self.inputs.extra_values[i])
+            self._results['out_file'].append(out_file)
+            self._results['compression'].append(_copy_any(fname, out_file))
+        return runtime
```

Comment:
We already have a DerivativesDataSink:

https://github.com/nipreps/dmriprep/blob/0a609b4c91c97b6b2bb0fad0f35ce2e7d15f44b3/dmriprep/interfaces/__init__.py#L13-L16

Unless there is some new functionality here, this can go away.

Reply:
yep!